include/execution.hpp

```diff
@@ -2171,7 +2171,222 @@ namespace std::execution {
+          while(_op_state != nullptr) {
+            __operation_base *__next = _op_state->__next_;
+            _op_state->__notify();
+            _op_state = __next;
+          }
+        }
+      };
+
+    template <class _SenderId, class _ReceiverId>
+      class __operation : public __operation_base {
+        using _Sender = __t<_SenderId>;
+        using _Receiver = __t<_ReceiverId>;
+
+        struct __on_stop_requested {
+          in_place_stop_source& __stop_source_;
+          void operator()() noexcept {
+            __stop_source_.request_stop();
+          }
+        };
+        using __on_stop = optional<typename stop_token_of_t<
+            env_of_t<_Receiver> &>::template callback_type<__on_stop_requested>>;
+
+        _Receiver __recvr_;
+        __on_stop __on_stop_{};
+        shared_ptr<__sh_state<_SenderId>> __shared_state_;
+
+      public:
+        __operation(_Receiver&& __rcvr,
+                    shared_ptr<__sh_state<_SenderId>> __shared_state)
+          : __operation_base()
+          , __recvr_((_Receiver&&)__rcvr)
+          , __shared_state_(move(__shared_state)) {
+        }
+
+        void __notify() noexcept override {
+          __on_stop_.reset();
+
+          std::visit([&](const auto& __tupl) noexcept -> void {
+            std::apply([&](auto __tag, const auto&... __args) noexcept -> void {
+              __tag((_Receiver&&) __recvr_, __args...);
+            }, __tupl);
+          }, __shared_state_->__data_);
+        }
+
+        friend void tag_invoke(start_t, __operation& __self) noexcept try {
+          __sh_state<_SenderId>* __shared_state = __self.__shared_state_.get();
+          atomic<void*>& __head = __shared_state->__head_;
+          void* const __completion_state = static_cast<void*>(__shared_state);
+          void* __old = __head.load(memory_order_acquire);
+
+          if (__old != __completion_state) {
+            __self.__on_stop_.emplace(
+                get_stop_token(get_env(__self.__recvr_)),
+                __on_stop_requested{__shared_state->__stop_source_});
+          }
+
+          do {
+            if (__old == __completion_state) {
+              __self.__notify();
+              return;
+            }
+            __self.__next_ = static_cast<__operation_base*>(__old);
+          } while (!__head.compare_exchange_weak(
+              __old, static_cast<void *>(&__self),
+              memory_order_acq_rel,
+              memory_order_acquire));
+
+          if (__old == nullptr) {
+            start(__shared_state->__op_state2_);
+          }
+        } catch (...) {
+          execution::set_error((_Receiver&&) __self.__recvr_, current_exception());
+        }
+      };
+
+    template <class _SenderId>
+      class __sender {
+        using _Sender = __t<_SenderId>;
+        using __sh_state = __sh_state<_SenderId>;
+        template <class _Receiver>
+          using __operation = __operation<_SenderId, __x<remove_cvref_t<_Receiver>>>;
+
+        _Sender __sndr_;
```

Comment:
In this case, I won't be able to forward sender queries.

Reply:
I'll move it into the shared state for now. Can we consider having sender environments for cases like this one?

Reply:
I agree that this is a problem, but I think it's a problem with the design of `split()` and not this implementation. `split()` needs to decide whether it supports move-only senders or the forwarding of sender queries. Given the current design, it can't do both.

One possibility is to move sender queries into an environment, as with receivers. Then we can copy the environment into the shared state and throw away the sender.